x-pack/plugins/infra/server/lib/alerting/metric_threshold/lib/evaluate_alert.ts

```diff
@@ -173,31 +198,55 @@ const getMetric: (
   }
 };
 
+interface DropPartialBucketOptions {
+  from: number;
+  to: number;
+  bucketSizeInMillis: number;
+}
+
+const dropPartialBuckets = ({ from, to, bucketSizeInMillis }: DropPartialBucketOptions) => (
+  row: {
+    key: string;
+    value: number;
+  } | null
+) => {
+  if (row == null) return null;
+  const timestamp = new Date(row.key).valueOf();
+  return timestamp >= from && timestamp + bucketSizeInMillis <= to;
+};
+
 const getValuesFromAggregations = (
   aggregations: Aggregation,
-  aggType: MetricExpressionParams['aggType']
+  aggType: MetricExpressionParams['aggType'],
+  dropPartialBucketsOptions: DropPartialBucketOptions
 ) => {
   try {
     const { buckets } = aggregations.aggregatedIntervals;
     if (!buckets.length) return null; // No Data state
     if (aggType === Aggregators.COUNT) {
-      return buckets.map((bucket) => ({
-        key: bucket.to_as_string,
-        value: bucket.doc_count,
-      }));
+      return buckets
+        .map((bucket) => ({
+          key: bucket.from_as_string,
+          value: bucket.doc_count,
+        }))
+        .filter(dropPartialBuckets(dropPartialBucketsOptions));
```

Comment:
I'm asking this here, but it applies to all places where you filter
Why are you checking all buckets? If I understood it correctly, the only bucket that could be wrong is the last one, wouldn't it be enough to check the last one and remove it if it has data outside the timerange?

Reply:
That would be an optimization, just running it as a `filter` on the array is more convenient and less complex.